### ISSUE TITLE
NullAway not properly checking method references

### ISSUE THREAD

**Issue by msridhar** [uber/NullAway#138]
Example:
```java
import javax.annotation.Nullable;
public class FunctionTest {
  interface Function<T, R> {
    R apply(T t);
  }
  static <R,T> R map(T t, Function<T, R> fun) {
    return fun.apply(t);
  }
  @Nullable
  static Object doSomething(String t) {
    return null;
  }
  static void test() {
    String ex = "hi";
    // BUG: should not be able to pass doSomething here
    map(ex, FunctionTest::doSomething);
  }
}
```
This code currently passes NullAway, but we should be reporting an error at the call to `map`.

**Comment by lazaroclapp** [comment]
Do we currently support something like `Function<T, @Nullable R>` that allows us to distinguish the case where a method reference can return null vs when it can't?
